pubsub/google/cloud/pubsub_v1/publisher/batch/thread.py

```diff
@@ -0,0 +1,385 @@
+        if len(self._messages) == 0:
+            return
+        response = self.client.api.publish(
+            self._topic,
+            self.messages,
+        )
+
+        # We got a response from Pub/Sub; denote that we are processing.
+        self._status = 'processing results'
+
+        # Sanity check: If the number of message IDs is not equal to the
+        # number of futures I have, then something went wrong.
+        if len(response.message_ids) != len(self._futures):
+            raise exceptions.PublishError(
+                'Some messages were not successfully published.',
+            )
+
+        # Iterate over the futures on the queue and return the response IDs.
+        # We are trusting that there is a 1:1 mapping, and raise an exception
+        # if not.
+        self._status = self.Status.SUCCESS
+        for message_id, fut in zip(response.message_ids, self._futures):
+            self.message_ids[hash(fut)] = message_id
+            fut._trigger()
+
+    def monitor(self):
+        """Commit this batch after sufficient time has elapsed.
+
+        This simply sleeps for ``self._settings.max_latency`` seconds,
+        and then calls commit unless the batch has already been committed.
+        """
+        # Note: This thread blocks; it is up to the calling code to call it
+        # in a separate thread.
+        #
+        # Sleep for however long we should be waiting.
+        time.sleep(self._settings.max_latency)
+
+        # If, in the intervening period, the batch started to be committed,
+        # then no-op at this point.
+        if self._status != self.Status.ACCEPTING_MESSAGES:
+            return
+
+        # Commit.
+        return self._commit()
+
+    def publish(self, message):
+        """Publish a single message.
+
+        Add the given message to this object; this will cause it to be
+        published once the batch either has enough messages or a sufficient
+        period of time has elapsed.
+
+        This method is called by :meth:`~.PublisherClient.publish`.
+
+        Args:
+            message (~.pubsub_v1.types.PubsubMessage): The Pub/Sub message.
+
+        Returns:
+            ~.pubsub_v1.publisher.batch.mp.Future: An object conforming to the
+                :class:`concurrent.futures.Future` interface.
+        """
+        # Coerce the type, just in case.
+        if not isinstance(message, types.PubsubMessage):
+            message = types.PubsubMessage(**message)
+
+        # Add the size to the running total of the size, so we know
+        # if future messages need to be rejected.
+        self._size += message.ByteSize()
+
+        # Store the actual message in the batch's message queue.
+        self._messages.append(message)
+
+        # Return a Future. That future needs to be aware of the status
+        # of this batch.
+        f = Future(self)
+        self._futures.append(f)
+        return f
+
+
+class Future(object):
```

Comment:
Please use the future class I spent so long creating in `google.cloud.core.future`. :)

(You may need to subclass it, but that's fine)